sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetReadSupport.scala

```diff
@@ -167,20 +168,33 @@ object ParquetReadSupport {
   }
 
   private def clipParquetType(
-      parquetType: Type, catalystType: DataType, caseSensitive: Boolean): Type = {
+      parquetType: Type,
+      catalystType: DataType,
+      caseSensitive: Boolean,
+      enableVectorizedReader: Boolean): Type = {
     catalystType match {
       case t: ArrayType if !isPrimitiveCatalystType(t.elementType) =>
         // Only clips array types with nested type as element type.
-        clipParquetListType(parquetType.asGroupType(), t.elementType, caseSensitive)
+        clipParquetListType(
+          parquetType.asGroupType(), t.elementType, caseSensitive, enableVectorizedReader)
 
       case t: MapType
         if !isPrimitiveCatalystType(t.keyType) ||
            !isPrimitiveCatalystType(t.valueType) =>
         // Only clips map types with nested key type or value type
-        clipParquetMapType(parquetType.asGroupType(), t.keyType, t.valueType, caseSensitive)
+        clipParquetMapType(
+          parquetType.asGroupType(), t.keyType, t.valueType, caseSensitive, enableVectorizedReader)
 
       case t: StructType =>
-        clipParquetGroup(parquetType.asGroupType(), t, caseSensitive)
+        clipParquetGroup(parquetType.asGroupType(), t, caseSensitive, enableVectorizedReader)
+
+      case t: DecimalType if enableVectorizedReader =>
+        val p = parquetType.asPrimitiveType().getDecimalMetadata()
+        if (t.precision == p.getPrecision && t.scale == p.getScale) {
+          parquetType
+        } else {
+          throw new UnsupportedOperationException("Schema evolution not supported.")
```

Comment:
Shall we tell that MR path is supported for the reading? I image that Spark apps might be able to read parquet, but once the reading adds a complex column, then it fails to read. Maybe confusing for end-users.

Reply:
If we add `complex` type columns like nested structure, it will succeed to read because vectorized reader is only enabled for atomic columns.

Reply:
Although this follows the previous exception style, we may add some detailed reason like `Parquet vectorized reader doesn't support schema evolution on decimal types. You may try to use `${PARQUET_VECTORIZED_READER_ENABLED.key}=false`. Does it sound more clear?

Reply:
yea, sounds good.